
## LabDS01 ##


The following dataset includes information about several countries. This information was collected by the world bank (https://data.worldbank.org/). Data corresponds to 2016.

`
dataFile="https://raw.githubusercontent.com/masterfloss/data/main/worlddata2016.xlsx"
df=pd.read_excel(dataFile)
`


# 1. Business Understanding

Purpose: Analyse data and extract knowledge

Observation: Country


In [36]:
#Importar bibliotecas
import pandas as pd
import numpy as np
from sklearn.decomposition import PCA
from sklearn import preprocessing
from sklearn.preprocessing import StandardScaler

In [37]:
#Importar dados
dataFile="https://raw.githubusercontent.com/masterfloss/data/main/worlddata2016.xlsx"
df_o=pd.read_excel(dataFile)
df_o.head()

Unnamed: 0 Country Name                                     Indicator Name  \
0           0  Afghanistan  Access to clean fuels and technologies for coo...   
1           1  Afghanistan            Access to electricity (% of population)   
2           2  Afghanistan  Access to electricity, rural (% of rural popul...   
3           3  Afghanistan  Access to electricity, urban (% of urban popul...   
4           4  Afghanistan  Account ownership at a financial institution o...   

       2016  
0  32.44000  
1  97.70000  
2  97.09936  
3  99.50000  
4       NaN

In [38]:
df_o.shape

(312480, 4)

# 2. Data Undertanding

* Discuss the data and purposes
* What are the main limitations?

R: 
Uma das principais situações é fato de haver 312480 linhas de informação; pelo que se pode reagrupar os dados

# 3. Data Preparation

* Converting data to numeric
* Missing Values Imputation
* Normalization
* Variable Dummification
* Data balancing
* Pivot date in order to convert the values "Indicator Name" into comuns 
* Remove all columns that have no values, df1[columns].sum()==0
* Remove all columns that have more than 210 observations, df1[columns].count()<210

In [64]:
#Fazer a codificação das variaveis em substituição no nome apresentado
df=pd.pivot_table(df_o, values='2016', index=['Country Name'], columns=['Indicator Name'])
df.head()

Indicator Name  ARI treatment (% of children under 5 taken to a health provider)  \
Country Name                                                                       
Afghanistan                                                   NaN                  
Albania                                                       NaN                  
Algeria                                                       NaN                  
American Samoa                                                NaN                  
Andorra                                                       NaN                  

Indicator Name  Access to clean fuels and technologies for cooking (% of population)  \
Country Name                                                                           
Afghanistan                                                 32.44                      
Albania                                                     77.42                      
Algeria                                                     92.62                      
American Samoa                                                NaN                      
Andorra                                                    100.00                      

Indicator Name  Access to electricity (% of population)  \
Country Name                                              
Afghanistan                                   97.700000   
Albania                                      100.000000   
Algeria                                       99.990051   
American Samoa                                      NaN   
Andorra                                      100.000000   

Indicator Name  Access to electricity, rural (% of rural population)  \
Country Name                                                           
Afghanistan                                             97.099360      
Albania                                                100.000000      
Algeria                                                 99.965142      
American Samoa                                                NaN      
Andorra                                                100.000000      

Indicator Name  Access to electricity, urban (% of urban population)  \
Country Name                                                           
Afghanistan                                                  99.5      
Albania                                                     100.0      
Algeria                                                     100.0      
American Samoa                                                NaN      
Andorra                                                     100.0      

Indicator Name  Adequacy of social insurance programs (% of total welfare of beneficiary households)  \
Country Name                                                                                           
Afghanistan                                                   NaN                                      
Albania                                                       NaN                                      
Algeria                                                       NaN                                      
American Samoa                                                NaN                                      
Andorra                                                       NaN                                      

Indicator Name  Adequacy of social protection and labor programs (% of total welfare of beneficiary households)  \
Country Name                                                                                                      
Afghanistan                                                   NaN                                                 
Albania                                                       NaN                                                 
Algeria                                                       NaN                                                 
American Samoa                                                NaN 

Pivot_table = cruzamento de variaveis

Do DF original, os nomes das variaveis passam a ser as colunas 

Só os dados de 2016 

In [53]:
df.shape

(217, 1328)

In [54]:
#Remover as colunas que não tenham valores
df_new=df
for columns in df_new:
    if(df_new[columns].sum==0):
        df_new=df_new.drop([columns], axis=1)

In [55]:
df_new

Indicator Name         ARI treatment (% of children under 5 taken to a health provider)  \
Country Name                                                                              
Afghanistan                                                          NaN                  
Albania                                                              NaN                  
Algeria                                                              NaN                  
American Samoa                                                       NaN                  
Andorra                                                              NaN                  
...                                                                  ...                  
Virgin Islands (U.S.)                                                NaN                  
West Bank and Gaza                                                   NaN                  
Yemen, Rep.                                                          NaN                  
Zambia                                                               NaN                  
Zimbabwe                                                             NaN                  

Indicator Name         Access to clean fuels and technologies for cooking (% of population)  \
Country Name                                                                                  
Afghanistan                                                        32.44                      
Albania                                                            77.42                      
Algeria                                                            92.62                      
American Samoa                                                       NaN                      
Andorra                                                           100.00                      
...                                                                  ...                      
Virgin Islands (U.S.)                                                NaN                      
West Bank and Gaza                                                   NaN                      
Yemen, Rep.                                                        64.93                      
Zambia                                                             16.43                      
Zimbabwe                                                           29.05                      

Indicator Name         Access to electricity (% of population)  \
Country Name                                                     
Afghanistan                                          97.700000   
Albania                                             100.000000   
Algeria                                              99.990051   
American Samoa                                             NaN   
Andorra                                             100.000000   
...                                                        ...   
Virgin Islands (U.S.)                               100.000000   
West Bank and Gaza                                   99.990334   
Yemen, Rep.                                          70.856285   
Zambia                                               35.307339   
Zimbabwe                                             39.923992   

Indicator Name         Access to electricity, rural (% of rural population)  \
Country Name                                                                  
Afghanistan                                                    97.099360      
Albania                                                       100.000000      
Algeria                                                        99.965142      
American Samoa                                                       NaN      
Andorra                                                       100.000000      
...                                                                  ...      
Virgin Islands (U.S.)                                         100.000000      
West Bank 

In [61]:
# "converter os "0"s em NaN
df_new=df_new.replace(0,np.nan)

In [62]:
#Remover as colunas com menos de 210 Observações (do exercicio)
for column1 in df_new:
    if df_new[column1].count()<210:
        df_new=df_new.drop([column1], axis=1)

In [65]:
df_new.shape

(217, 18)

Com a redução aplicada no ponto anterior, passou-se de 1328 colunas para apenas 18 colunas

In [66]:
df_new.isnull().any()

Indicator Name
Access to electricity (% of population)                                  True
Access to electricity, urban (% of urban population)                     True
Agricultural land (sq. km)                                               True
Capture fisheries production (metric tons)                               True
Land area (sq. km)                                                       True
People using at least basic drinking water services (% of population)    True
Population density (people per sq. km of land area)                      True
Population growth (annual %)                                             True
Population, total                                                        True
Secure Internet servers                                                  True
Secure Internet servers (per 1 million people)                           True
Surface area (sq. km)                                                    True
Terrestrial and marine protected areas (% of tota

In [69]:
#Remover linhas com NaN
df_new=df_new.dropna()
df_new.shape

(194, 18)

In [88]:
ColList=[]
for col1 in df_new:
    ColList.append(col1)
ColList

['Access to electricity (% of population)',
 'Access to electricity, urban (% of urban population)',
 'Agricultural land (sq. km)',
 'Capture fisheries production (metric tons)',
 'Land area (sq. km)',
 'People using at least basic drinking water services (% of population)',
 'Population density (people per sq. km of land area)',
 'Population growth (annual %)',
 'Population, total',
 'Secure Internet servers',
 'Secure Internet servers (per 1 million people)',
 'Surface area (sq. km)',
 'Terrestrial and marine protected areas (% of total territorial area)',
 'Terrestrial protected areas (% of total land area)',
 'Total fisheries production (metric tons)',
 'Urban population',
 'Urban population (% of total population)',
 'Urban population growth (annual %)']

In [108]:
#criar uma lsita com os nomes dos "Countries"
CountryList=df_new.index
CountryList

Index(['Afghanistan', 'Albania', 'Algeria', 'Angola', 'Antigua and Barbuda',
       'Argentina', 'Armenia', 'Aruba', 'Australia', 'Austria',
       ...
       'United States', 'Uruguay', 'Uzbekistan', 'Vanuatu', 'Venezuela, RB',
       'Vietnam', 'Virgin Islands (U.S.)', 'Yemen, Rep.', 'Zambia',
       'Zimbabwe'],
      dtype='object', name='Country Name', length=194)

---------------------------------------------------------------

* Nesta fase já temos a lista de index e também alguma limpeza dos dados iniciais

In [133]:
from sklearn.preprocessing import StandardScaler
dfnSt=preprocessing.StandardScaler().fit_transform(df_new)

In [134]:
pcaModel=PCA(n_components=4)

In [135]:
features_pca=pcaModel.fit_transform(dfnSt)

In [136]:
scores=pcaModel.components_.T

In [137]:
dfscores=pd.DataFrame(scores)

In [138]:
dfscores['Variaveis']=ColList

In [140]:
dfscores

0         1  \
Variaveis                                                                
Access to electricity (% of population)             0.126534 -0.423074   
Access to electricity, urban (% of urban popula...  0.120927 -0.391063   
Agricultural land (sq. km)                          0.369825  0.110467   
Capture fisheries production (metric tons)          0.378317  0.096403   
Land area (sq. km)                                  0.334811  0.084307   
People using at least basic drinking water serv...  0.118009 -0.425011   
Population density (people per sq. km of land a... -0.007381 -0.082952   
Population growth (annual %)                       -0.090174  0.347838   
Population, total                                   0.349933  0.121130   
Secure Internet servers                             0.182080 -0.032199   
Secure Internet servers (per 1 million people)      0.046430 -0.182299   
Surface area (sq. km)                               0.333916  0.082737   
Terrestrial and marine protected areas (% of to...  0.024235 -0.036873   
Terrestrial protected areas (% of total land area) -0.006096 -0.090642   
Total fisheries production (metric tons)            0.343376  0.104295   
Urban population                                    0.392524  0.102933   
Urban population (% of total population)            0.090004 -0.294900   
Urban population growth (annual %)                 -0.072746  0.396441   

                                                           2         3  
Variaveis                                                               
Access to electricity (% of population)            -0.103078 -0.019245  
Access to electricity, urban (% of urban popula... -0.109958 -0.035152  
Agricultural land (sq. km)                          0.098813  0.174949  
Capture fisheries production (metric tons)         -0.108186 -0.214703  
Land area (sq. km)                                  0.152950  0.355029  
People using at least basic drinking water serv... -0.066091 -0.004360  
Population density (people per sq. km of land a... -0.013500  0.004825  
Population growth (annual %)                        0.069487  0.114551  
Population, total                                  -0.133022 -0.263400  
Secure Internet servers                             0.270674  0.359424  
Secure Internet servers (per 1 million people)      0.168457  0.192709  
Surface area (sq. km)                               0.154206  0.359874  
Terrestrial and marine protected areas (% of to...  0.631166 -0.304242  
Terrestrial protected areas (% of total land area)  0.578269 -0.387322  
Total fisheries production (metric tons)           -0.153494 -0.332157  
Urban population                                   -0.085476 -0.202946  
Urban population (% of total population)            0.113295  0.149898  
Urban population growth (annual %)                  0.058811  0.026484

In [ ]:
#Gerar um heat map...

In [142]:
def selectColor(val):
    if (val > 0.3 or val < -0.3):
        color='red'
    else:
        color='black'
    return 'color: %s' % color
df_cor=dfscores.style.applymap(selectColor)
df_cor

Que concluções tirar deste mapa?

JLeal, 26 novembro, revisto +/-